docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,126 @@
+
+ 
+### Empty Profile
+
+A device profile can begin life “empty” - meaning that it has no device resources or device commands.  Once associated to any object (device, event, reading) then, since device resource or device command additions can be made, the empty profile can gain resources or commands.  You can never remove device resources or device commands once the profile is associated to another object.
+
+The general properties of a device profile, with the exception of name, are optional when creating an empty profile.
+
+All profiles – even empty profiles – must have a name.                                                                                 
+
+### Deleting Profiles
+
+A profile can only be deleted from metadata when the profile is not associated to any other EdgeX object to include device, provision watcher, event or reading.
+
+### Association Changes
+
+Today, a device profile is associated to a device via its name (the ProfileName on the Device struct).  The profile can also associated to events and readings (via the ProfileName in the Event and Reading objects).
+
+Profiles can be associated to more than one device.  Profiles can be associated to multiple events/readings.
+
+Rules about change of a device profile association have never been explicitly defined.  While the section above outline what can/cannot be changed in the profile and under what circumstances, this section outlines whether a device, event or reading can alter the associated device profile and under what circumstances.
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system.  There is too much risk that an existing event or reading also contains the device profile reference – causing complexity in knowing what information is in the event/reading.
+
+A provision watcher also has an association to a device profile.  However, the device profile is only used to know the type of device to create in case of provisioning.  Therefore, if a device profile changes, it does not impact the provision watcher.
+
+A device profile cannot be deleted while it is associated to a provision watcher
+
+Change of the device profile on an event or reading is never allowed.  Events and readings are immutable.
+
+### Applicability
+
+This document dictates the rules around allowed change to device profiles (and associations).  All EdgeX services will abide by these rules (implementing appropriate checks and validation as necessary to enforce the rules) when this ADR is approved to include:
+- service APIs
+- user interfaces (like the Edge GUI)
+- client tools (like the EdgeX CLI)
+
+## Proposed Design
+
+The following metadata API changes are suggested as a means to implement the device profile rules of this ADR
+
+- Block Device DELETE when there is any associated event/reading exists.
+- Block Profile (Upload) PUT when there is any associated device/event/reading exists.
```

Comment:
The swagger for a device profile put says that it "allows updates to an existing device profile". If so, why is this being blocked? I'm also not sure what `(Upload)` means in this context?

Reply:
This is where we are trying to align on rules.  Updates of a profile should have only been allowed when a profile was not associated to anything.  This was not checked in the current system (and why the UI was able to bypass)

Upload means to upload a file (versus post or put JSON).

Reply:
I thought in our last discussion that enforcing profile rules via `PUT` was hard, which is why we decided to add the helper APIs?